From c93b960ae7be5492f699d7420721b6d9975cab69 Mon Sep 17 00:00:00 2001
From: Hisashi Nakamura <hisashi.nakamura.ak@renesas.com>
Date: Wed, 12 Jun 2013 15:09:02 +0900
Subject: [PATCH 485/715] ARM: shmobile: r8a7790: Add power management support
 for RGX

Signed-off-by: Hisashi Nakamura <hisashi.nakamura.ak@renesas.com>
(cherry picked from commit 61f4bf69cb7f46508d01e884359b424af8bc6165)

Signed-off-by: Ryo Kataoka <ryo.kataoka.wt@renesas.com>
---
 arch/arm/mach-shmobile/include/mach/r8a7790.h |    2 ++
 arch/arm/mach-shmobile/pm-r8a7790.c           |    7 +++++++
 arch/arm/mach-shmobile/setup-r8a7790.c        |    4 ++++
 3 files changed, 13 insertions(+)

diff --git a/arch/arm/mach-shmobile/include/mach/r8a7790.h b/arch/arm/mach-shmobile/include/mach/r8a7790.h
index bbc1940..2a96b5b 100644
--- a/arch/arm/mach-shmobile/include/mach/r8a7790.h
+++ b/arch/arm/mach-shmobile/include/mach/r8a7790.h
@@ -33,6 +33,8 @@ extern int r8a7790_sysc_power_down(struct r8a7790_pm_ch *r8a7790_ch);
 extern int r8a7790_sysc_power_up(struct r8a7790_pm_ch *r8a7790_ch);
 
 #ifdef CONFIG_PM
+extern struct r8a7790_pm_domain r8a7790_rgx;
+
 extern void r8a7790_init_pm_domain(struct r8a7790_pm_domain *r8a7790_pd);
 extern void r8a7790_add_device_to_domain(struct r8a7790_pm_domain *r8a7790_pd,
 					 struct platform_device *pdev);
diff --git a/arch/arm/mach-shmobile/pm-r8a7790.c b/arch/arm/mach-shmobile/pm-r8a7790.c
index cdd9747..0eed6bc 100644
--- a/arch/arm/mach-shmobile/pm-r8a7790.c
+++ b/arch/arm/mach-shmobile/pm-r8a7790.c
@@ -214,6 +214,13 @@ void r8a7790_add_device_to_domain(struct r8a7790_pm_domain *r8a7790_pd,
 		pm_clk_add(dev, NULL);
 }
 
+struct r8a7790_pm_domain r8a7790_rgx = {
+	.ch = {
+		.chan_offs = 0xc0, /* PWRSR2 .. PWRER2 */
+		.isr_bit = 20, /* RGX */
+	}
+};
+
 #endif /* CONFIG_PM */
 
 void __init r8a7790_pm_init(void)
diff --git a/arch/arm/mach-shmobile/setup-r8a7790.c b/arch/arm/mach-shmobile/setup-r8a7790.c
index b84aff2..6df55ef 100644
--- a/arch/arm/mach-shmobile/setup-r8a7790.c
+++ b/arch/arm/mach-shmobile/setup-r8a7790.c
@@ -165,6 +165,8 @@ void __init r8a7790_add_standard_devices(void)
 {
 	r8a7790_pm_init();
 
+	r8a7790_init_pm_domain(&r8a7790_rgx);
+
 	r8a7790_register_scif(SCIFA0);
 	r8a7790_register_scif(SCIFA1);
 	r8a7790_register_scif(SCIFB0);
@@ -174,6 +176,8 @@ void __init r8a7790_add_standard_devices(void)
 	r8a7790_register_scif(SCIF0);
 	r8a7790_register_scif(SCIF1);
 	r8a7790_register_irqc(0);
+
+	r8a7790_add_device_to_domain(&r8a7790_rgx, &powervr_device);
 }
 
 void __init r8a7790_timer_init(void)
-- 
1.7.10.4

